### PR TITLE
Remote control and albums identification improvements

### PR DESCRIPTION
**Pull request by ddv239**
Make the Album artist be listed as the album's Primary artist and distinguish albums by the combination of (Album name, Album artist).
Ensure remote is notified of the correct playback progress in all scenarios attempted.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -52,3 +53,7 @@ android {
         abortOnError false
     }
 }
+
+dependencies {
+    implementation 'com.android.support:support-annotations:27.0.2'
```
Comment (adrian-bl) [inline build.gradle]: Why do we need this? I'm not a big fan of the support library and try to avoid it.

File: src/ch/blinkenlights/android/medialibrary/MediaScanner.java [hunk 2]
```diff
@@ -487,13 +487,20 @@ private boolean rpcInspectFile(File file) {
...
+
+			long albumartistId = MediaLibrary.hash63(albumartist);
 
-			// Overwrite albumId with a hash that included the parent dir if set in preferences
+			// I have several albums all called "Greatest Hits" with different album artists
+			String albumIdString = album + "\n" + albumartist;
+			// Include the parent dir in the hash if set in preferences
 			if (prefs.groupAlbumsByFolder) {
-				albumId = MediaLibrary.hash63(album + "\n" + file.getParent());
+				albumIdString += "\n" + file.getParent();
```
Comment (adrian-bl) [inline src/ch/blinkenlights/android/medialibrary/MediaScanner.java]: What is the intend of these changes? Making the AlbumArtist the 'primary' artist-source of an Album?
I think this should rather be a config option where the user can choose if he prefers artist-related actions on the artist or album-artist. These changes would probably cause issue with various corner cases (and would require a full library flush)

File: src/ch/blinkenlights/android/vanilla/RemoteControlImplLp.java [hunk 3]
```diff
@@ -111,6 +116,17 @@ public void reloadPreference() {
 		mShowCover = -1;
 	}
 
+	// HACK: This static is only needed as I found no other way to force correct progress display in the following scenarios:
+	// I.  1. Vanilla is started after a "Force stop"
+	//     2. BlueSoleil 10.0.497.0 is connected
```
Comment (adrian-bl) [inline src/ch/blinkenlights/android/vanilla/RemoteControlImplLp.java]: What is BlueSoleil and why do we need to implement a workaround for it?
Also, these changes would probably be better off in a separate pull request.

**Comment by Adonai** [comment]
AFAIK Adrian still builds Vanilla with Ant, so you have to mirror deps to the build.xml

File: build.gradle (hunk 1 above)
Comment (Adonai) [reply build.gradle]: My guess is that it's just annotations part of it for making code look cleaner (e.g. `@NonNull` or `@Nullable` etc. JSR-305)

Make it "compileOnly" so It doesn't add anything to runtime.

File: src/ch/blinkenlights/android/vanilla/RemoteControlImplLp.java (hunk 3 above)
Comment (ddv239) [reply src/ch/blinkenlights/android/vanilla/RemoteControlImplLp.java]: OK, I'll split this PR into separate ones.

File: build.gradle (hunk 1 above)
Comment (ddv239) [reply build.gradle]: Yes, that is my intention for supporting the annotations.
I did not know about "compileOnly", will do. I agree it will be nice to make sure there is no extra unneeded runtime dependency.

File: src/ch/blinkenlights/android/vanilla/RemoteControlImplLp.java (hunk 3 above)
Comment (ddv239) [reply src/ch/blinkenlights/android/vanilla/RemoteControlImplLp.java]: BlueSoleil an AVRCP-capable Bluetooth receiver for Windows (http://www.bluesoleil.com/products/S0001201005190001.html).
I think it makes perfect sense to make sure that Vanilla's interoperation with the Remote Control code is done in such a way that as many publicly available receivers as possible work as perfectly as possible, so BlueSoleil is one good test case for this goal. It is not clear which of the components in the pair "Remote Control code in Android + BlueSoleil AVRCP code" is guilty of the scenario I described in the comments for which this workaround is intended, and it is possible that one of those components makes some extra assumptions not guaranteed by the AVRCP specifications, but it is not realistic to expect one of them to fix this, so implementing this (fairly benign and harmless) workaround is no doubt a reasonable thing for us to do.

File: src/ch/blinkenlights/android/medialibrary/MediaScanner.java (hunk 2 above)
Comment (ddv239) [reply src/ch/blinkenlights/android/medialibrary/MediaScanner.java]: > What is the intend of these changes? Making the AlbumArtist the 'primary' artist-source of an Album?

Yes.

> I think this should rather be a config option where the user can choose if he prefers artist-related actions on the artist or album-artist. 

I understand your point, but let's be realistic, can you imagine any real-life scenario where a user would prefer to use the "artist" field value that is different form the "album artist" value, to identify the album's artist and display that as the property of the album? I cannot imagine such an example, but if you can present one, I would like to learn about it.

Here is my rationale for why it would not make sense. The cases I can imagine where both "Artist" and "Album artist" fields are present in the song and are different are the following:
- the album is "an album of Artist1" and may include some of "featured artists" on some of the tracks.
- the album is "a compilation of a specific genre/type of music" performed by various artists. Then the "Album artist" value _should_ be "VA" or "Various Artists" and "Artist" values of individual songs could vary.
- the album is "an album of songs originally performed/authored by a band/composer/poet". In this case, the appropriate value for the "Album artist" field is most probably the name of that original author and individual songs' "Artist" values would vary.

As you can easily see, in all of those cases above, using the "artist" value for album's property always only creates confusion and never correctly achieves the goal of identifying the set of all the songs in this album as a proper album. This consideration makes me believe that this piece of logic should not be configurable but "Album artist" should always be used for identifying the sets of songs as albums. By the way, Google Play Music does this unconditionally too and that seems very reasonable.

I have nothing against a config option when it makes sense, but if having this option "off" would never make any sense for any real-life user, then we better not have the "off" option.

> These changes would probably cause issue with various corner cases 

Considering the recently introduced calculation of "Album artist" based on "Artist" when the former is absent, I don't really anticipate any such "corner cases".

> (and would require a full library flush)

Yes, I am not sure what is the right way to achieve this but I am sure there is an existing mechanism for such flushes and we should use it here.

**Comment by ddv239** [comment]
I tried building the current Vanilla sources from the master branch with Ant and found that the current version of the Android SDK intentionally dropped support for it (see https://stackoverflow.com/questions/42912824/the-ant-folder-is-suddenly-missing-from-android-sdk-did-google-remove-it). Considering this, I would say "let's remove support for building Vanilla with Ant too". What do you guys think?

**Comment by ddv239** [comment]
As suggested before, I am separating this PR into 2 parts: Gradle upgrade and Remote Control.

File: build.gradle (hunk 1 above)
Comment (Adonai) [reply build.gradle]: Welcome aboard, I see you're catching up pretty quickly ;)

**Comment by Adonai** [comment]
> Considering this, I would say "let's remove support for building Vanilla with Ant too". What do you guys think?

I'm all for it, but we'll have to rethink F-Droid build (should look at fdroiddata repo) and fix it in case it still uses it.

**Comment by ddv239** [comment]
BTW, despite the above, Ant build still works successfully after my Gradle upgrade if I replace the Android SDK's Tools directory with its older version that contauns Ant support.
